pkg/agent/flowexporter/connections/connections.go

```diff
@@ -0,0 +1,134 @@
+// Copyright 2020 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package connections
+
+import (
+	"sync"
+	"time"
+
+	"k8s.io/klog"
+
+	"github.com/vmware-tanzu/antrea/pkg/agent/flowexporter"
+	"github.com/vmware-tanzu/antrea/pkg/agent/interfacestore"
+	"github.com/vmware-tanzu/antrea/pkg/agent/openflow"
+)
+
+var _ ConnectionStore = new(connectionStore)
+
+type ConnectionStore interface {
+	Run(stopCh <-chan struct{})
+}
+
+type connectionStore struct {
+	connections map[flowexporter.ConnectionKey]flowexporter.Connection // Add 5-tuple as string array
+	connDumper  ConnTrackDumper
+	ifaceStore  interfacestore.InterfaceStore
+	mutex       sync.Mutex
+}
+
+func NewConnectionStore(ctDumper ConnTrackDumper, ifaceStore interfacestore.InterfaceStore) *connectionStore {
+	return &connectionStore{
+		connections: make(map[flowexporter.ConnectionKey]flowexporter.Connection),
+		connDumper:  ctDumper,
+		ifaceStore:  ifaceStore,
+	}
+}
+
+// Run polls the connTrackDumper module periodically to get connections. These connections are used
+// to build connection store. If there is an error in poll cycle, we break the loop and exit the routine.
+func (cs *connectionStore) Run(stopCh <-chan struct{}) {
+	klog.Infof("Starting conntrack polling")
+
+	ticker := time.NewTicker(flowexporter.PollInterval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-stopCh:
+			break
+		case <-ticker.C:
+			_, err := cs.poll()
+			if err != nil {
+				klog.Errorf("Error during conntrack poll cycle: %v", err)
+				break
```

Comment:
is breaking the right thing to do? Is there a possibility for a transient error that will be resolved in the next few iterations?

Reply:
I wanted to save cycles for the case, where there are consistent failures when polling conntrack connections. I can see only one possibility of transient error, when trying to get netlink socket through Dial function. Not sure how often it is possible--any comments?

Reply:
It's a bit weird I think to give up after one failed iteration, with no possibility of resuming polling at a later time. I think we should either remove the break statement, or have some backoff mechanism (increase interval up to 1 or 2 minutes).

Reply:
Makes sense. Removed break. Added TODO for backoff mechanism by increasing poll interval. In addition, we could have a timeout to exit from routine if there are consistent errors.